### PR TITLE
client,cmd,daemon,release: better version info

### PR DESCRIPTION
**Pull request by chipaca**
Resurrect @zyga's old #1222 and further tweak it:
```
snap:  2.0.24
snapd: 2.0.24/16/Ubuntu 16.04 LTS/classic
```

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -195,12 +195,22 @@ func (client *Client) ServerVersion() (string, error) {
...
-	version := sysInfo.Version
-	if version == "" {
-		version = "unknown"
+	serverVersion := sysInfo.Version
+	if serverVersion == "" {
+		serverVersion = "unknown"
 	}
 
-	return fmt.Sprintf("%s (series %s)", version, sysInfo.Series), nil
+	what := "classic"
```
Comment (zyga) [inline client/client.go]: s/what/variant/?

File: client/client.go [hunk 2]
```diff
@@ -195,12 +195,22 @@ func (client *Client) ServerVersion() (string, error) {
...
+	if !sysInfo.OnClassic {
+		what = "unspecified"
+		// what = machine info (e.g. RPi3 etc)
```
Comment (zyga) [inline client/client.go]: ah, hmm, so that would be the model assertion here, right?

File: client/client.go [hunk 3]
```diff
@@ -195,12 +195,22 @@ func (client *Client) ServerVersion() (string, error) {
...
+	}
+
+	// where := "localhost" ... :-)
```
Comment (zyga) [inline client/client.go]: I see what you did there ;-)

File: client/client.go [hunk 4]
```diff
@@ -244,10 +254,21 @@ func IsTwoFactorError(err error) bool {
 	return e.Kind == ErrorKindTwoFactorFailed || e.Kind == ErrorKindTwoFactorRequired
 }
 
+// OSRelease contains information about the system extracted from /etc/os-release.
+type OSRelease struct {
+	ID         string `json:"id"`
+	Name       string `json:"name"`
+	VersionID  string `json:"version-id,omitempty"`
+	Codename   string `json:"codename,omitempty"`
```
Comment (zyga) [inline client/client.go]: Codename is an ubuntu-specific extension. AFAIR we still set it to "xenial" if it is missing. That shouldn't be happening.

File: client/client.go [hunk 5]
```diff
@@ -244,10 +254,21 @@ func IsTwoFactorError(err error) bool {
...
+	PrettyName string `json:"pretty-name,omitempty"`
+}
+
 // SysInfo holds system information
 type SysInfo struct {
-	Series  string `json:"series,omitempty"`
-	Version string `json:"version,omitempty"`
+	Series    string    `json:"series,omitempty"`
+	Version   string    `json:"version,omitempty"`
+	OSRelease OSRelease `json:"os-release"`
```
Comment (zyga) [inline client/client.go]: Nice!

File: client/client.go [hunk 6]
```diff
@@ -244,10 +254,21 @@ func IsTwoFactorError(err error) bool {
 	return e.Kind == ErrorKindTwoFactorFailed || e.Kind == ErrorKindTwoFactorRequired
 }
 
+// OSRelease contains information about the system extracted from /etc/os-release.
+type OSRelease struct {
+	ID         string `json:"id"`
```
Comment (zyga) [inline client/client.go]: I'm somewhat on edge about this but I'd love to consider using the original `/etc/os-release` spelling of those. So `ID`, `VERSION_ID`, `PRETTY_NAME` etc.

File: release/release.go [hunk 7]
```diff
@@ -48,7 +51,7 @@ func (os *OS) ForceDevMode() bool {
 		return false
 
 	case "elementary":
-		switch os.Release {
+		switch os.VersionID {
```
Comment (zyga) [inline release/release.go]: Thank you :-)

File: release/release.go [hunk 8]
```diff
@@ -69,31 +72,39 @@ var osReleasePath = "/etc/os-release"
...
+			osRelease.PrettyName = v
 		}
 	}
-	if osRelease.Codename == "" {
-		osRelease.Codename = "xenial"
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("cannot read os-release: %s", err)
+	}
+
+	if osRelease.Name == "ubuntu" && osRelease.Codename == "" {
```
Comment (zyga) [inline release/release.go]: This should test `id`, not `name`. I also wonder why we even have this bit?

File: client/client.go (hunk 4 above)
Comment (chipaca) [inline client/client.go]: not any more we don't :-)

**Comment by niemeyer** [comment]
I would still like to see a cleaner format, as proposed on the original branch:
```
snap   2.0.1
snapd  2.0.2
series 16
ubuntu 16.04
```

**Comment by niemeyer** [comment]
Miss-click, sorry..

**Comment by mvo5** [comment]
The spread test: `2016/06/27 17:31:07 Error executing linode:ubuntu-16.04-32-grub:tests/upgrade/basic :` is failing in `snapdver=$(snap --version|grep "snapd ")`.

**Comment by niemeyer** [comment]
From yesterday on Telegram:

Gustavo Niemeyer, [27.06.16 14:06]
@Chipaca I also don't understand the point about arch.. surely it has some sort of versioning?

Gustavo Niemeyer, [27.06.16 14:12]
@Chipaca About the colons, the suggestion there is without them.. as a list of pairs that looks fine

Gustavo Niemeyer, [27.06.16 14:12]
s/fine/nice

Gustavo Niemeyer, [27.06.16 14:14]
@Chipaca Arch does have a version: https://www.archlinux.org/releng/releases/

Gustavo Niemeyer, [27.06.16 14:14]
@Chipaca If they're not publishing that somewhere we can consume, I'd just put "unknown" there

Gustavo Niemeyer, [27.06.16 14:15]
archlinux   unknown

Gustavo Niemeyer, [27.06.16 14:16]
(I'd say archlinux rather than arch)

**Pull request merged**